docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,592 @@
+  - Cosign verifier should be able to operate a loose ('All keys are equally trusted and ONLY one is requred to be used in validation') or strict (ALL keys are trusted and ALL keys must be used in validation) enforcement.
+
+### How does Cosign CLI verification work today?
+
+- Image manifest pushed to same repository as subject (tag schema: `sha256-abc123.sig`)
+- Image manifest is updated every time new signature is added. Each layer is a different signature (possibly from different signing keys)
+```json
+# Sample Cosign OCI Image Manifest
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "config": {
+    "mediaType": "application/vnd.oci.image.config.v1+json",
+    "size": 451,
+    "digest": "sha256:bbacdc3dc4b33c868c6c27192cb1d1bda619c1454abb57ebea806eac13250d3c"
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
+      "size": 254,
+      "digest": "sha256:75a45fb043a8abfe52f256cb1a424d8808e886a3f40886228cdfa9b8d75c805d",
+      "annotations": {
+        "dev.cosignproject.cosign/signature": "MEQCIGoZazNt1/eEBc08AmcqXY+yNQQJY2ziZPiIopDjHnU5AiAUgeWBy90GgqDdnWpxd1b3lbqzgr2UV2ZyR59eDOyxVg=="
+      }
+    },
+    {
+      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
+      "size": 254,
+      "digest": "sha256:75a45fb043a8abfe52f256cb1a424d8808e886a3f40886228cdfa9b8d75c805d",
+      "annotations": {
+        "dev.cosignproject.cosign/signature": "MEUCIBVcGGZR6pvJ2SQ8KI0TZY02SLFSSOFJuJph2Qcm8cWZAiEArL+BF5BGRGxELVBeJtAUy8i7L1uuBeUT9fF6TrFYVLQ="
+      }
+    },
+    {
+      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
+      "size": 254,
+      "digest": "sha256:75a45fb043a8abfe52f256cb1a424d8808e886a3f40886228cdfa9b8d75c805d",
+      "annotations": {
+        "dev.cosignproject.cosign/signature": "MEQCIEIsK5yr1kGS1dgRmbudT55oghADdyAOfUwlnayPKAmmAiBTdZkhgHkZdw11tullz7HNOg5O1r5U/j4m2RE39b92Vg=="
+      }
+    }
+  ]
+}
+```
+- Image signature succeeds if AT LEAST one of the signatures verifies true with the corresponding key
+  - Validation occurs on each signature in the manifest in parallel against the SAME key
+  - Cosign only supports validation against a SINGLE key at a time
+
+### Trust Policy
+Cosign verifier should support multiple trust policies based on the KeyManagementProviders (KMS) enabled and the desired verification scenario. Please refer to [this](#user-scenarios) section for common user scenarios. At a high level users must be able to have:
+  - multiple KMS `inline` key resources (each `inline` will have a single key)
+  - multiple keys defined in a single AKV KMS
+  - multiple KMS `inline` certificate resources (each `inline` may have a single cert or a cert chain)
+  - multiple certificates definine in a single AKV KMS
+  - mix of keys + certificates in a single AKV KMS
+  - a way to specify specific keys/certificates within a KMS by name and version
```

Comment:
We need a universal standard provider agnostic on how to uniquely identify certificates/keys. For inline provider I propose we store the content in the map with a keyed value of `<namespace>/<type>/<name>` and for AKV we can do `<namespace>/<type>/<name>/<certificate name>/?<version>?`